hack/generator/pkg/codegen/storage/storage_type_factory.go

```diff
@@ -0,0 +1,465 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package storage
+
+import (
+	"fmt"
+
+	"github.com/Azure/k8s-infra/hack/generator/pkg/astmodel"
+	"github.com/pkg/errors"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog/v2"
+)
+
+// StorageTypeFactory is used to create storage types for a specific service
+type StorageTypeFactory struct {
+	service                    string                     // Name of the service we're handling (used mostly for logging)
+	types                      astmodel.Types             // All the types for this service
+	propertyConversions        []propertyConversion       // Conversion rules to use for properties when creating storage variants
+	pendingStorageConversion   astmodel.TypeNameQueue     // Queue of types that need storage variants created for them
+	pendingConversionInjection astmodel.TypeNameQueue     // Queue of types that need conversion functions injected
+	pendingMarkAsHubVersion    astmodel.TypeNameQueue     // Queue of types that need to be flagged as the hub storage version
+	idFactory                  astmodel.IdentifierFactory // Factory for creating identifiers
+	storageConverter           astmodel.TypeVisitor       // a cached type visitor used to create storage variants
+	propertyConverter          astmodel.TypeVisitor       // a cached type visitor used to simplify property types
+	functionInjector           astmodel.TypeVisitor       // a cached type visitor used to inject functions into definitions
+	resourceHubMarker          astmodel.TypeVisitor       // a cached type visitor used to mark resources as Storage Versions
+
+	// Map of conversion links for creating our conversion graph
+	// (Can't use PackageReferences as keys, so keyed by the full package path)
+	conversionMap map[string]astmodel.PackageReference
```

Comment:
Is this right? There are other places we use `PackageReference`s as keys (like in `PackageReferenceSet`). Didn't you mention that all the implementations have value semantics?

Reply:
yeah IIRC we can use `PackageReference` as keys and it works.